### DISCUSSION TITLE
Using Strongswan Client for Android

### DISCUSSION THREAD

**Discussion by Misha0736** [strongswan/strongswan#2498] [General]
Hello.

We are using Strongswan Client for Android for connect VPN IKEv2 Server. The VPN-server is running on NGFW UserGate. Authentication type is IKEv2 EAP (Login/Password). Everything works in this configuration.

To increase security, we decided to add a second authorization factor. The connection is also successful on most android devices. But an error occurs on some devices. I am attaching the log of the problematic device. The connection is interrupted on the second retransmit, followed by an error "unable to schedule job for execution in 3919 ms". At the same time, the application shows that the connection attempt is continuing
[Problem_device.log](https://github.com/user-attachments/files/17393129/Problem_device.log)

What could be the problem? I suspect the problem is the short timeout of the response from the server, because without the second factor, these devices connect without problems

**Comment by tobiasbrunner** [comment]
What you see in the log is due to a permission problem. The app requires the ability to call [`setExactAndAllowWhileIdle()`](https://developer.android.com/reference/android/app/AlarmManager#setExactAndAllowWhileIdle(int,%20long,%20android.app.PendingIntent)) to schedule tasks with a timeout greater than 3 seconds. However, that requires the `SCHEDULE_EXACT_ALARM` permission or basically adding the app to the system's power whitelist (the app requests users to do so if it's not yet the case and the check has not been disabled in the app settings). At least that's the case on a standard Android image. Depending on the vendor, it's possible that additional permissions have to be granted to the app in order to call that function, so check if there are any permission related to power usage that you can grant to the app in the system settings.

**Comment by kisaa** [comment]
Is this issue resolved by any way? I have Android 13 / MIUI 14.0.7, and faced with same problem. Server logs are clear:

2026-04-02T11:11:02.105711+10:00 swan charon: 10[IKE] sending DPD request
2026-04-02T11:11:02.106749+10:00 swan charon: 10[IKE] queueing IKE_DPD task
2026-04-02T11:11:02.106888+10:00 swan charon: 10[IKE] activating new tasks
2026-04-02T11:11:02.107024+10:00 swan charon: 10[IKE]   activating IKE_DPD task
2026-04-02T11:11:02.107128+10:00 swan charon: 10[ENC] generating INFORMATIONAL request 61 [ ]
2026-04-02T11:11:02.107241+10:00 swan charon: 10[NET] sending packet: from X.X.X.X[4500] to 185.156.1.123[9319] (80 bytes)
2026-04-02T11:11:02.204917+10:00 swan charon: 16[NET] received packet: from 185.156.1.123[9319] to X.X.X.X[4500] (80 bytes)
2026-04-02T11:11:02.205208+10:00 swan charon: 16[ENC] parsed INFORMATIONAL response 61 [ ]
2026-04-02T11:11:02.205397+10:00 swan charon: 16[IKE] activating new tasks
2026-04-02T11:11:02.205571+10:00 swan charon: 16[IKE] nothing to initiate

but client stuck the same way. MIUI has no option to add such permission to any app. Battery mode is already set to "No limits".

**Reply by tobiasbrunner** [reply]
What do you mean? That log doesn't show any issues.

**Reply by kisaa** [reply]
You are right, server log is fine. But client log is:

Apr  2 16:20:13 00[DMN] +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
Apr  2 16:20:13 00[DMN] Starting IKE service (strongSwan 6.0.3, Android 13 - TKQ1.221114.001 test-keys/2024-09-01, 220333QNY - Redmi/rain_ru/Xiaomi, Linux 4.19.157-perf-g98ae9330c4dd, aarch64, org.strongswan.android)
Apr  2 16:20:13 00[LIB] providers loaded by OpenSSL: default legacy
Apr  2 16:20:13 00[LIB] loaded plugins: androidbridge charon android-log socket-default openssl nonce pkcs1 pem x509 xcbc kdf revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls
Apr  2 16:20:13 00[JOB] spawning 16 worker threads
Apr  2 16:20:14 07[CFG] loaded user certificate 'CN=xxxxxx, O=xxxxxxxxxxxx' and private key
Apr  2 16:20:14 07[CFG] loaded CA certificate 'CN=xxxxxxxxxxxxxxx'
Apr  2 16:20:14 07[IKE] initiating IKE_SA android[2] to X.X.X.X
Apr  2 16:20:14 07[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Apr  2 16:20:14 07[NET] sending packet: from 10.87.140.230[37761] to X.X.X.X[500] (948 bytes)
Apr  2 16:20:16 12[IKE] retransmit 1 of request with message ID 0
Apr  2 16:20:16 12[NET] sending packet: from 10.87.140.230[37761] to X.X.X.X[500] (948 bytes)
Apr  2 16:20:18 08[IKE] retransmit 2 of request with message ID 0
Apr  2 16:20:18 08[NET] sending packet: from 10.87.140.230[37761] to X.X.X.X[500] (948 bytes)
Apr  2 16:20:18 08[JOB] unable to schedule job for execution in 3919 ms

Just as topic starter wrote - " At the same time, the application shows that the connection attempt is continuing"

**Reply by tobiasbrunner** [reply]
This is an initial connection attempt. That doesn't match your server log (where the client responded to a DPD request). And it indicates some fundamental network issue.
